src/Build/BackEnd/Node/MSBuildClient.cs

```diff
@@ -0,0 +1,661 @@
+
+        #region Message pump
+        /// <summary>
+        /// The packet factory.
+        /// </summary>
+        private readonly NodePacketFactory _packetFactory;
+
+        /// <summary>
+        /// Shared read buffer.
+        /// </summary>
+        private SharedReadBuffer _sharedReadBuffer;
+
+        /// <summary>
+        /// The queue of packets we have received but which have not yet been processed.
+        /// </summary>
+        private readonly ConcurrentQueue<INodePacket> _receivedPacketsQueue;
+
+        /// <summary>
+        /// Set when packet pump receive packets and put them to <see cref="_receivedPacketsQueue"/>.
+        /// </summary>
+        private readonly AutoResetEvent _packetReceivedEvent;
+
+        /// <summary>
+        /// Set when the asynchronous packet pump should terminate.
+        /// </summary>
+        private ManualResetEvent _terminatePacketPumpEvent;
+
+        /// <summary>
+        /// Set when we packet pump enexpectedly shutdown (due to connection problems or becuase of desearilization issues).
+        /// </summary>
+        private readonly ManualResetEvent _packetPumpShutdownEvent;
+
+        /// <summary>
+        /// The thread which runs the asynchronous packet pump
+        /// </summary>
+        private Thread? _packetPump;
+        #endregion
+
+        // TODO: work on eleminating extra parameters or making them more clearly described at least.
+        /// <summary>
+        /// Public constructor with parameters.
+        /// </summary>
+        /// <param name="msbuildLocation">Location of msbuild dll or exe.</param>
+        /// <param name="exeLocation">Location of executable file to launch the server process.
+        /// That should be either dotnet.exe or MSBuild.exe location.</param>
+        /// <param name="dllLocation">Location of dll file to launch the server process if needed.
+        /// Empty if executable is msbuild.exe and not empty if dotnet.exe.</param>
+        public MSBuildClient(string msbuildLocation, string exeLocation, string dllLocation)
+        {
+            ServerEnvironmentVariables = new();
+            _exitResult = new();
+
+            // dll & exe locations
+            _msBuildLocation = msbuildLocation;
+            _exeLocation = exeLocation;
+            _dllLocation = dllLocation;
+
+            // Client <-> Server communication stream
+            _handshake = GetHandshake();
+            _pipeName = NamedPipeUtil.GetPipeNameOrPath("MSBuildServer-" + _handshake.ComputeHash());
+            _nodeStream = new NamedPipeClientStream(".", _pipeName, PipeDirection.InOut, PipeOptions.Asynchronous
+#if FEATURE_PIPEOPTIONS_CURRENTUSERONLY
+                                                                         | PipeOptions.CurrentUserOnly
+#endif
+            );
+
+            _packetMemoryStream = new MemoryStream();
+            _binaryWriter = new BinaryWriter(_packetMemoryStream);
+
+            // Packet pump block
```

Comment:
NOT: word block is confusing (prohibit vs block-of-code). I recommend something like: Initialize packet pump